operator/pkg/ciliumendpointslice/reconciler.go

```diff
@@ -29,71 +35,174 @@ func newReconciler(client clientset.CiliumV2alpha1Interface, cesMgr operations)
+	} else if exists && desiredCEPsNumber == 0 {
+		return r.reconcileCESDelete(cesObj)
+	} else { //!exist && desiredCEPsNumber == 0 => no op
+		return nil
+	}
+}
 
-	if len(ces.Endpoints) == 0 {
-		// The CES is empty, delete ces information from cache rather than creating
-		// empty CES.
-		r.cesManager.deleteCESFromCache(cesToCreate)
-		return
+// Create a new CES in api-server.
+func (r *reconciler) reconcileCESCreate(cesName CESName) (err error) {
+	log.WithFields(logrus.Fields{
+		logfields.CESName: cesName,
+	}).Debug("Reconciling CES Create.")
+	ceps := r.cesManager.getCEPinCES(cesName)
+	if operatorOption.Config.EnableMetrics {
+		metrics.CiliumEndpointsChangeCount.WithLabelValues(metrics.LabelValueCEPInsert).Observe(float64(len(ceps)))
+	}
+	newCES := &cilium_v2a1.CiliumEndpointSlice{
+		TypeMeta: meta_v1.TypeMeta{
+			Kind:       "CiliumEndpointSlice",
+			APIVersion: cilium_v2.SchemeGroupVersion.String(),
+		},
+		ObjectMeta: meta_v1.ObjectMeta{
+			Name: string(cesName),
+		},
+		Endpoints: make([]cilium_v2a1.CoreCiliumEndpoint, 0, len(ceps)),
+	}
+
+	cesData := r.cesManager.getCESData(cesName)
+	newCES.Namespace = cesData.ns
+
+	for _, cepName := range ceps {
+		ccep := getCoreEndpointFromStore(cepName)
+		if ccep != nil {
+			newCES.Endpoints = append(newCES.Endpoints, *ccep)
+		}
 	}
 
 	// Call the client API, to Create CES
-	if retCES, err = r.client.CiliumEndpointSlices().Create(
-		context.TODO(), ces, metav1.CreateOptions{}); err != nil {
+	if _, err = r.client.CiliumEndpointSlices().Create(
+		context.TODO(), newCES, meta_v1.CreateOptions{}); err != nil {
 		log.WithError(err).WithFields(logrus.Fields{
-			logfields.CESName: cesToCreate,
+			logfields.CESName: newCES.Name,
 		}).Info("Unable to create CiliumEndpointSlice in k8s-apiserver")
-		return
 	}
-	// Update local datastore CES with latest object metadata values.
-	r.cesManager.updateCESInCache(retCES, false)
 	return
 }
 
 // Update an existing CES
-func (r *reconciler) reconcileCESUpdate(cesToUpdate string) (err error) {
-	var retCES, ces *cilium_v2.CiliumEndpointSlice
+func (r *reconciler) reconcileCESUpdate(cesName CESName, cesObj *cilium_v2a1.CiliumEndpointSlice) (err error) {
+	log.WithFields(logrus.Fields{
+		logfields.CESName: cesName,
+	}).Debug("Reconciling CES Update.")
+	updatedCES := cesObj.DeepCopy()
+	cesEqual := true
+	cepInserted := 0
+	cepRemoved := 0
+	cepUpdated := 0
 
-	// Before syncing with ApiServer, get list of removed CEPs
-	remCEPs := r.cesManager.getRemovedCEPs(cesToUpdate)
-	// Get the copy of CES from the cesManager
-	if ces, err = r.cesManager.getCESCopyFromCache(cesToUpdate); err != nil {
-		return
+	// Iterate through existing CEPs in the CES
+	// remove stale ones
+	// update outdated
+	// Iterate through expected CEPs in the CES
+	// insert missing
+	ceps := r.cesManager.getCEPinCES(cesName)
+	desiredEndpoints := make([]cilium_v2a1.CoreCiliumEndpoint, 0, len(ceps))
+	desiredObjects := make(map[CEPName]*cilium_v2a1.CoreCiliumEndpoint)
```

Comment:
It's a little difficult to follow `desiredEndpoints` and `desiredObjects`.
Consider renaming `desiredEndpoints` to `updatedEndpoints`, and `desiredObjects` to `desiredEndpoints`.

Reply:
Simplified a code a little bit. It may reshuffle the CEPs on update now but it's much easier to track